packages/quantic/force-app/main/default/lwc/headlessLoader/headlessLoader.js

```diff
@@ -0,0 +1,150 @@
+import HeadlessPath from '@salesforce/resourceUrl/coveoheadless';
+import AtomicPath from '@salesforce/resourceUrl/atomicutils';
+// @ts-ignore	
+import { loadScript } from 'lightning/platformResourceLoader';
+
+const DEBOUNCE_DELAY = 200;
+let timeout;
+
+/**
+ * Debounces a function execution.
+ * @param {Function} func The function for which to delay execution.
+ * @param {Number} wait The time to delay in milliseconds.
+ */
+const debounce = (func, wait) => {
+  return function executedFunction(...args) {
+    const later = () => {
+      clearTimeout(timeout);
+      func(...args);
+    };
+    clearTimeout(timeout);
+    // eslint-disable-next-line @lwc/lwc/no-async-operation
+    timeout = setTimeout(later, wait);
+  };
+};
+
+/**
+ * Cancels the delayed search query.
+ */
+const cancelInitialSearch = () => {
+  if (timeout) {
+    clearTimeout(timeout);
+    timeout = undefined;
+  }
+}
+
+/**
+ * Dispatches search request.
+ */
+const executeInitialSearch = debounce(() => {
+  window.coveoHeadless.engine.dispatch(
+    CoveoHeadless.SearchActions.executeSearch(
+      CoveoHeadless.AnalyticsActions.logInterfaceLoad()
+    )
+  );
+}, DEBOUNCE_DELAY);
```

Comment:
Why is a debounce needed if `executeInitialSearch` is called only once `areAllComponentsInitialized` is true?

Reply:
When components are in separate sections of a page (ex: searchbox in the page header and the resultlist in the page content section), there's a race condition that occurs. Sometimes the searchbox is initialized before the resultlist constructor is called. The debounce gives a time buffer so the resultlist can cancel the request. With this 200ms timer value i've yet to notice another instance of two initial requests being sent.

Reply:
I would * not * recommend to rely on this. It is asking for trouble, from my experience ;) 


The initialization process is not something you want to "mostly works most of the time", its one the place where it's really worth it to do some over-engineering to really nail it and be really confident.

Either the delay is entirely un-necessary as you explain in your document, so you could remove it, and then you are all good.

Or it is necessary/there are corner case where it would be needed. 

And if that's the case, it needs to be a "solid" central orchestration logic, and must not rely on any timeout/debounce/luck.

Reply:
I agree. with this the "worst case scenario" is simply that an initial query might be incorrect and a second query is made once the late component arrives so it doesn't create a disaster explosion either. But yes I agree it needs to be improved. Would you be willing to accept it as being an upcoming improvement?

Btw @olamothe as you might remember we talked a couple weeks ago about a quantic error i was getting with locker service. If my investigation is correct, the current state of quantic doesn't work with locker as redux and the proxies used during parent-child initialization don't mix well. So while this solution isn't perfect yet, I think it still leaves things in a better spot than currently.
Thoughts?

Reply:
>  If my investigation is correct, the current state of quantic doesn't work with locker as redux and the proxies used during parent-child initialization don't mix well. So while this solution isn't perfect yet, I think it still leaves things in a better spot than currently.
> Thoughts?

It does leave it in a better state I guess. 

But I would encourage you to go deep into understanding what is the underlying problem with locker service + redux, to isolate the actual problem per say, and to do any change required in headless if needed to make it work, or if there is any issue that could be reported back to Salesforce developers.

I think it will pay off in the long run.

Reply:
That's good advice. What we know so far is that redux dispatch calls Object.freeze() on a proxy object which doesn't work as desired. It would be interesting to see what salesforce has to say about that. We'll look into it. 👍
Thanks for your input!